### DISCUSSION TITLE
카카오 정상예약 가능 여부

### DISCUSSION THREAD

**Discussion by PREMI3R** [kimytsc/covid-rest-vaccine-macro#62] [General]
안녕하세요 좋은 자료 개발 및 공개해주셔서 감사합니다.

안내에 따라 준비과정을 완료하고 동작도중 아래와 같은 이슈가 있어

기능상 문제인지 단순 딜레이에 따른 오류인지 여쭤보려 남깁니다

지정한 좌표에서 새로고침중 한 병원에서 2개의 잔여백신이 발생하였고

발생 -> 예약실패 -> 새로고침 -> 새로고침 -> 새로고침 -> 발생 -> 예약실패

![fidd](https://user-images.githubusercontent.com/89119016/129835283-253d99c1-cf84-4485-a8eb-00defcf49a9e.png)

통신내역

![change](https://user-images.githubusercontent.com/89119016/129835285-3409496a-50cd-436c-9632-8e9414a2fab8.png)

각각의 통신에서 잔여백신 2-> 1로 정상적으로 줄어듦

이와 같은 경우는 단순히 갱신등이 늦게되어서 발생하는 문제일까요?

아니면 예약상 어떠한 변경점이 있어서 예약진행이 안된건지 궁금합니다

다시 한번 노고에 감사드립니다

**Comment by kimytsc** [comment]
해당 통신내역과 본문의 내용을 가지고 분석을 해보자면 다음과 같을 것으로 추정됩니다.

1. `hidden request 1740`: 병원 리스트 확인, `11353228`병원의 총 잔여백신 수량 존재 확인
2. `request 1741`: `11353228`병원 정보에서 화이자 2개 확인
3. `request 1743`: 화이자 예약시도, 실패(=403 응답, 응답코드 및 메세지 확인 필요)
4. `request 1747`: 병원 리스트 확인, `11353228`병원의 총 잔여백신 수량이 `2` 또는 `0`
5. `request 1748`: 병원 리스트 확인, `11353228`병원의 총 잔여백신 수량이 `2` 또는 `0`
6. `request 1749`: 병원 리스트 확인, `11353228`병원의 총 잔여백신 수량 변경 확인(`2` 또는 `0` -> `1`)
7. `request 1750`: `11353228`병원 정보에서 화이자 1개 확인
8. `request 1751`: 화이자 예약시도, 실패(=403 응답, 응답코드 및 메세지 확인 필요)

여기서 `request 1747` ~ `request 1749`의 `새로고침` 과정에서 `11353228`병원의 잔여백신이 0개 또는 2개가 남아 있을 수 있습니다.

0개가 남았다고 가정할 경우, `request 1743`의 예약 시도 실패(=이미 잔여백신이 없어졌을 경우 403 응답)와 연결고리가 성립합니다.
그리고 `request 1750`에서 다시 잔여백신이 갱신되었고 `request 1751`에서 또다시 실패했을 가능성이 있습니다.

2개가 남아있었다고 가정할 경우, AZ때문에 트래픽을 조절하기 위해 1분간 동일한 병원의 동일한 잔여백신 수량에 대해서는 예약 시도하지 않도록 설정되어 있어 새로고침만 되었을 것입니다. 그리고 다시 `request 1750`에서 1분이 지나 `request 1751 잔여백신 예약 시도`를 진행하였을 것으로 추정됩니다. 이 경우, `request 1743`, `request 1750`의 예약 시도 실패는 `백신접종 예약내역이 있거나 이미 접종을 하신 경우 잔여백신 접종 신청이 불가합니다.`의 사유가 되었을 가능성이 높습니다.

다만, 새로고침 간격을 짧게(delay: default 1000ms) 했을 가능성이 높아 `request 1747` ~ `request 1749` 과정에서 잔여백신은 0개가 떴을 것 확률이 높아 보이며, 예약 시도할 당시 잔여백신은 이미 마감되었을 가능성이 있습니다.

최근 사전 백신예약을 통한 실패 사례가 나오고 있어 fiddler에서 `request 1743`, `request 1751`의 응답코드 및 메세지 또는 화면에 노출된 예약시도 상태 메세지의 확인이 있어야 보다 정확한 판단이 가능할 것 같습니다.
